src/markdown_exec/formatters/jupyter.py

```diff
@@ -0,0 +1,253 @@
+# specific runner when needed.
+def _run_jupyter(
+    language: str,
+    start_kernel_kwargs: dict[str, Any] | None,
+    code: str,
+    returncode: int | None = None,
+    session: str | None = None,
+    id: str | None = None,  # noqa: A002, ARG001
+    **extra: str,
+) -> str:
+    """Execute code using a Jupyter kernel."""
+    is_named_session = session is not None and session != ""
+
+    # Get kernel manager and client
+    km, kc = _get_kernel_manager_and_client(language, session, start_kernel_kwargs)
+
+    kc.wait_for_ready()
+    # Send code for execution
+    msg_id = kc.execute(code, store_history=is_named_session)
+
+    # Collect output
+    outputs = []
+    error_outputs = []
+    execution_completed = False
+
+    while not execution_completed:
+        try:
+            # Check for shell messages (execution replies)
+            shell_msg = kc.get_shell_msg(timeout=0.1)
+            if shell_msg["parent_header"].get("msg_id") == msg_id:
+                execution_completed = True
+
+                if shell_msg["content"]["status"] == "error":
+                    if (
+                        shell_msg["content"]["traceback"] is not None
+                        and len(shell_msg["content"]["traceback"]) > 0
+                    ):
+                        if shell_msg["content"]["traceback"][0] == "Stack trace:":
+                            error_outputs.extend(
+                                shell_msg["content"]["traceback"][1:],
+                            )
+                        else:
+                            error_outputs.extend(shell_msg["content"]["traceback"])
+                    break
+                # note: abort is deprecated, but some older kernels may still send it
+                # it doesn't send error info, however - so we'll need to populate a fake abort error instead
+                if shell_msg["content"]["status"] == "abort":
+                    error_outputs.append("AbortError: Execution aborted")
+                    break
+        except queue.Empty:
+            pass
+
+        try:
+            # nested while here so we consume all iopub messages for this execution before moving on.
+            # this lets us consume all pending messages after execution has completed before exiting the loop
+            while True:
+                # Check for iopub messages (output)
+                iopub_msg = kc.get_iopub_msg(timeout=0.1)
+                if iopub_msg["parent_header"].get("msg_id") != msg_id:
+                    continue
+
+                msg_type = iopub_msg["header"]["msg_type"]
+                content = iopub_msg["content"]
+
+                if msg_type == "stream":
+                    outputs.append(content["text"])
+                elif msg_type in ("execute_result", "display_data"):
+                    # TODO: these can contain other MIME types (including HTML) - do we want to support them?
```

Comment:
answer: Markdown Exec supports Markdown and HTML outputs, so it sounds like a good idea to also support HTML output. I'm not sure what that implies with Jupyter though.